### PR TITLE
init

### PR DESCRIPTION
**Pull request by SirJohanot**

### CODE REVIEW

File: src/main/java/com/epam/ball/dao/DataReader.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    public DataReader() {
+    }
+
+    public List<String> read(String path) throws DataException {
+        List<String> result = new ArrayList<>();
+        try {
+            File file = new File(path);
+            BufferedReader bufferedReader = new BufferedReader(new FileReader(file));
+            String temp;
+            while ((temp=bufferedReader.readLine())!=null)
```
Comment (FilippStankevichTR) [inline src/main/java/com/epam/ball/dao/DataReader.java]: String line= bufferedReader.readLine(); 
 while ((temp=bufferedReader.readLine())!=null) {
     result.add(line);
     line= bufferedReader.readLine(); 
}

File: src/main/java/com/epam/ball/dao/DataReader.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+                result.add(temp);
+        } catch (FileNotFoundException e) {
+            throw new DataException("Could not find file by that path");
```
Comment (FilippStankevichTR) [inline src/main/java/com/epam/ball/dao/DataReader.java]: throw new DataException("Could not find file by that path", e);

File: src/main/java/com/epam/ball/dao/DataException.java [hunk 3]
```diff
@@ -0,0 +1,12 @@
+package com.epam.ball.dao;
+
+public class DataException extends Exception{
+
+    public DataException(){
+        super();
+    }
+
+    public DataException(String message){
```
Comment (FilippStankevichTR) [inline src/main/java/com/epam/ball/dao/DataException.java]: public DataException(String message, e)

File: src/main/java/com/epam/ball/entity/Ball.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        return radius;
+    }
+
+    public void setRadius(double radius) {
+        this.radius = radius;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
```
Comment (FilippStankevichTR) [inline src/main/java/com/epam/ball/entity/Ball.java]: if (this == o) {
 return true;
}

File: src/main/java/com/epam/ball/entity/Ball.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+public class Ball {
+
+    private Point center;
+    private double radius;
+
+    public Ball() {
+        center = new Point();
+    }
+
+    public Ball(double radius, Point center) {
```
Comment (FilippStankevichTR) [inline src/main/java/com/epam/ball/entity/Ball.java:11]: Point center, double radius

File: src/main/java/com/epam/ball/logic/BallCalculator.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+        return result;
+    }
+
+    public double calculateVolume(Ball ball) {
+        double result = 4.0 / 3.0 * Math.PI * Math.pow(ball.getRadius(), 3);
+        return result;
+    }
+
+    public double calculateTheRatioOfVolumesDividedByCoordinatePlane(Ball ball, CoordinatePlane plane) {
+        if (!isCrossingTheCoordinatePlane(ball, plane)) return 0;
```
Comment (FilippStankevichTR) [inline src/main/java/com/epam/ball/logic/BallCalculator.java]: if (!isCrossingTheCoordinatePlane(ball, plane)) {
   return 0;
}

File: src/main/java/com/epam/ball/logic/BallCreator.java [hunk 7]
```diff
@@ -0,0 +1,19 @@
+package com.epam.ball.logic;
+
+import com.epam.ball.entity.Ball;
+import com.epam.ball.entity.Point;
+
+public class BallCreator {
+
```
Comment (FilippStankevichTR) [inline src/main/java/com/epam/ball/logic/BallCreator.java:10]: private final static String DELEMITER = " ";

File: src/main/java/com/epam/ball/logic/BallValidator.java [hunk 8]
```diff
@@ -0,0 +1,15 @@
+package com.epam.ball.logic;
+
+import java.util.regex.Pattern;
+
+public class BallValidator {
+
+    private static final String BALL_LINE_REGEX="\\d+\\.\\d+ (-)?\\d+\\.\\d+ (-)?\\d+\\.\\d+ (-)?\\d+\\.\\d+";
```
Comment (FilippStankevichTR) [inline src/main/java/com/epam/ball/logic/BallValidator.java]: \\d+\\.\\d+ (-)?{4}

File: src/main/java/com/epam/ball/entity/Point.java [hunk 9]
```diff
@@ -0,0 +1,60 @@
+package com.epam.ball.entity;
+
+import java.util.Objects;
+
+public class Point {
+
+    private double xCoordinate;
```
Comment (FilippStankevichTR) [inline src/main/java/com/epam/ball/entity/Point.java]: final

File: src/main/java/com/epam/ball/logic/Director.java [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+import com.epam.ball.dao.DataException;
+import com.epam.ball.dao.DataReader;
+import com.epam.ball.entity.Ball;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class Director {
+
+    private DataReader reader;
```
Comment (FilippStankevichTR) [inline src/main/java/com/epam/ball/logic/Director.java]: final

File: src/main/java/com/epam/ball/logic/Director.java [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+    public Director(DataReader reader, BallValidator validator, BallCreator creator) {
+        this.reader = reader;
+        this.validator = validator;
+        this.creator = creator;
+    }
+
+    public List<Ball> read(String path) throws DataException {
+        List<Ball> balls=new ArrayList<>();
+        for (String line:reader.read(path)){
+            if (validator.isValidBallLine(line)){
```
Comment (FilippStankevichTR) [inline src/main/java/com/epam/ball/logic/Director.java]: isValidLine

File: src/main/java/com/epam/ball/logic/Director.java [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+                balls.add(creator.create(line));
```
Comment (FilippStankevichTR) [inline src/main/java/com/epam/ball/logic/Director.java]: Ball ball = creator.create(line);

File: src/test/java/com/epam/ball/logic/BallCalculatorTest.java [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+import com.epam.ball.entity.Ball;
+import com.epam.ball.entity.CoordinatePlane;
+import com.epam.ball.entity.Point;
+import org.junit.Assert;
+import org.junit.Test;
+
+public class BallCalculatorTest {
+
+    @Test
+    public void testShouldCalculateSurfaceAreaWhenBallIsValid() {
```
Comment (FilippStankevichTR) [inline src/test/java/com/epam/ball/logic/BallCalculatorTest.java]: testCalculateSurfaceAreaShouldCalculateWhenBallIValid

File: src/test/java/com/epam/ball/logic/BallCreatorTest.java [hunk 14]
```diff
@@ -0,0 +1,41 @@
+package com.epam.ball.logic;
+
+import com.epam.ball.entity.Ball;
+import com.epam.ball.entity.Point;
+import org.junit.Assert;
+import org.junit.Test;
+
+public class BallCreatorTest {
+
```
Comment (FilippStankevichTR) [inline src/test/java/com/epam/ball/logic/BallCreatorTest.java:9]: private static fina String VALID_LINE = "4.0 1.0 2.0 3.0";

Comment (FilippStankevichTR) [inline src/test/java/com/epam/ball/logic/BallCreatorTest.java:9]: private final  BallCreator creator=new BallCreator();

File: src/test/java/com/epam/ball/logic/DataReaderTest.java [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+    @Test
+    public void testShouldReadStringsWhenFileHasOneString() throws DataException {
+        //given
+        DataReader reader=new DataReader();
+        //when
+        List<String> strings=new ArrayList<>();
+        strings.add("1.0 2.0 3.0 4.0");
+        List<String> result=reader.read("src/test/resources/oneLineTxtFile");
+        //then
+        Assert.assertEquals(strings, result);
```
Comment (FilippStankevichTR) [inline src/test/java/com/epam/ball/logic/DataReaderTest.java]: Arrays.asList("1.0 2.0 3.0 4.0");

File: src/test/java/com/epam/ball/logic/DataReaderTest.java [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+        //given
+        DataReader reader=new DataReader();
+        //when
+        List<String> strings=new ArrayList<>();
+        strings.add("1.0 2.0 3.0 4.0");
+        strings.add("line");
+        strings.add("12345");
+        List<String> result=reader.read("src/test/resources/multipleLinesTxtFile");
+        //then
+        Assert.assertEquals(strings, result);
```
Comment (FilippStankevichTR) [inline src/test/java/com/epam/ball/logic/DataReaderTest.java]: List<String> expectedLines = Arrays.asList("1.0 2.0 3.0 4.0", "line", "12345");
  Assert.assertEquals(expectedLines  , result);

File: src/test/java/com/epam/ball/logic/DirectorTest.java [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class DirectorTest {
+
+    @Test
+    public void testShouldReadLinesWhenLinesAreValid() throws DataException {
+        //given
+        DataReader reader= Mockito.mock(DataReader.class);
```
Comment (FilippStankevichTR) [inline src/test/java/com/epam/ball/logic/DirectorTest.java]: DataReader reader= Mockito.mock(DataReader.class); 
Mockito.when(reader.read(path)).thenReturn(lines);

File: src/test/java/com/epam/ball/logic/DirectorTest.java [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+import com.epam.ball.entity.Point;
+import org.junit.Assert;
+import org.junit.Test;
+import org.mockito.Mockito;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class DirectorTest {
+
```
Comment (FilippStankevichTR) [inline src/test/java/com/epam/ball/logic/DirectorTest.java:15]: private final static String TEST_PATH = "path";

**Review by FilippStankevichTR — CHANGES_REQUESTED** [review]

File: src/main/java/com/epam/ball/dao/DataReader.java [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+            BufferedReader bufferedReader = new BufferedReader(new FileReader(file));
+            String line = bufferedReader.readLine();
+            while (line != null) {
+                result.add(line);
+                line = bufferedReader.readLine();
+            }
+        } catch (Exception e) {
+            DataException dataException = new DataException(e.getMessage(), e);
+            LOGGER.throwing(dataException);
+            throw dataException;
```
Comment (ingenmech) [inline src/main/java/com/epam/ball/dao/DataReader.java]: in one line throw new DataException(e.getMessage(), e);
no need log if you throw new exception

block finally for close your reader

File: src/main/java/com/epam/ball/entity/Ball.java [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+
+import java.util.Objects;
+
+public class Ball {
+
+    private Point center;
+    private double radius;
+
+    public Ball() {
+        center = new Point();
```
Comment (ingenmech) [inline src/main/java/com/epam/ball/entity/Ball.java]: private Point center = new Point() ; init in one line

File: src/main/java/com/epam/ball/entity/Point.java [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+import java.util.Objects;
+
+public class Point {
+
+    private final double xCoordinate;
+    private final double yCoordinate;
+    private final double zCoordinate;
+
+    public Point() {
+        xCoordinate = yCoordinate = zCoordinate = 0;
```
Comment (ingenmech) [inline src/main/java/com/epam/ball/entity/Point.java:10]: no need to init, They have been inited by default

File: src/main/java/com/epam/ball/logic/BallCalculator.java [hunk 22]
```diff
@@ -0,0 +1,71 @@
...
+        LOGGER.info("Started calculating the ratio of volumes of " + ball.toString() + " divided by " + plane.toString() + " plane");
+        if (!isCrossingTheCoordinatePlane(ball, plane)) {
+            return 0;
+        }
+        Point center = ball.getCenter();
+        double distanceFromCenterToPlane = distanceFromPlane(plane, center);
+        double radius = ball.getRadius();
+        double lesserHemisphereRadius = Math.sqrt(Math.pow(radius, 2) - Math.pow(distanceFromCenterToPlane, 2));
+        double lesserHemisphereVolume = 2.0 / 3.0 * Math.PI * Math.pow(lesserHemisphereRadius, 3);
+        double ratio = lesserHemisphereVolume / (calculateVolume(ball) - lesserHemisphereVolume);
```
Comment (ingenmech) [inline src/main/java/com/epam/ball/logic/BallCalculator.java:50]: you can write in one line : return lesserHemisphereVolume / (calculateVolume(ball) - lesserHemisphereVolume);

File: src/main/java/com/epam/ball/logic/BallCalculator.java [hunk 23]
```diff
@@ -0,0 +1,71 @@
...
+    public boolean isAValidBall(Ball ball) {
+        return ball.getRadius() > 0;
+    }
+
+    public boolean isCrossingTheCoordinatePlane(Ball ball, CoordinatePlane plane) {
+        LOGGER.info("Started calculating whether " + ball.toString() + " is crossing the " + plane.toString() + " plane");
+        Point center = ball.getCenter();
+        double coordinate = distanceFromPlane(plane, center);
+        double radius = ball.getRadius();
+        boolean isCrossing = Math.abs(coordinate) <= radius;
```
Comment (ingenmech) [inline src/main/java/com/epam/ball/logic/BallCalculator.java:64]: in one line no need variable, for all code

File: src/main/java/com/epam/ball/logic/BallCreator.java [hunk 24]
```diff
@@ -0,0 +1,29 @@
...
+import org.apache.logging.log4j.Logger;
+
+
+public class BallCreator {
+
+    private static final Logger LOGGER = LogManager.getLogger(BallCreator.class);
+
+    private static final String DELIMITER = " ";
+
+    public BallCreator() {
```
Comment (ingenmech) [inline src/main/java/com/epam/ball/logic/BallCreator.java]: no need empty constructors for all codes

File: src/main/java/com/epam/ball/logic/Director.java [hunk 25]
```diff
@@ -0,0 +1,41 @@
...
+        LOGGER.info("Started reading balls from file: " + path);
+        List<Ball> balls = new ArrayList<>();
+        try {
+            for (String line : reader.read(path)) {
+                if (validator.isValidLine(line)) {
+                    Ball ball = creator.create(line);
+                    balls.add(ball);
+                }
+            }
+        } catch (DataException e) {
```
Comment (ingenmech) [inline src/main/java/com/epam/ball/logic/Director.java]: why block catch is empty

File: src/main/java/com/epam/ball/repository/CenterXCoordinateBallComparator.java [hunk 26]
```diff
@@ -0,0 +1,17 @@
+package com.epam.ball.repository;
+
+import com.epam.ball.entity.BallObservable;
+import com.epam.ball.entity.Point;
+
+import java.util.Comparator;
+
+public class CenterXCoordinateBallComparator implements Comparator<BallObservable> {
+    @Override
+    public int compare(BallObservable o1, BallObservable o2) {
```
Comment (ingenmech) [inline src/main/java/com/epam/ball/repository/CenterXCoordinateBallComparator.java]: naming firstObserver, secondObserver

File: src/main/java/com/epam/ball/repository/CenterXCoordinateBallComparator.java [hunk 27]
```diff
@@ -0,0 +1,17 @@
...
+        Point center1 = o1.getCenter();
+        Point center2 = o2.getCenter();
+        double xCoordinate1 = center1.getXCoordinate();
+        double xCoordinate2 = center2.getXCoordinate();
```
Comment (ingenmech) [inline src/main/java/com/epam/ball/repository/CenterXCoordinateBallComparator.java]: naming

**Review by ingenmech — CHANGES_REQUESTED** [review]

**Pull request merged**